lib/whenParser.js

```diff
@@ -0,0 +1,158 @@
+"use strict";
+
+const convert = require("./convert");
+const utils = require("./utils");
+
+// Mostly a copy of ./parser.js/parseSchemaString, slightly modified.
+function parseSchemaString(value) {
+    let schemaString = value;
+    // No need to parse if if it isn't config, but just a string.
+    if (schemaString.indexOf(':') == -1 && schemaString.indexOf(',') == -1) {
+        return schemaString;
+    }
+
+    let type = null;
+    let schema = {};
+    // If we have a type, parse it.
+    if (schemaString.indexOf(':') > -1) {
+        let typeParts = schemaString.split(":", 2);
+
+        type = typeParts[0];
+        schema = {
+            "@type": type,
+        };
+        // Reset the string to parse the rest.
+        schemaString = typeParts[1];
+    }
+
+    // Parse any further config.
+    if (schemaString.length > 0) {
+        for (let valuePart of schemaString.split(",")) {
+            const parts = valuePart.split("=", 2);
+
+            let key = parts[0].trim();
+
+            let value = parts[1];
+
+            if (value) {
+                value = value.trim();
+            }
+            else {
+                value = null;
+            }
+
+            value = convert(type, key, value);
+
+            schema[key] = value;
+        }
+    }
+
+    return schema;
+}
+
+const whenParser = {
+    parse(value, engine) {
+        let condition = value.condition ? value.condition : null;
+        let options = {};
+
+        if (value.options) {
+            options = whenParser.parseOptions(value.options, engine);
+        }
+
+        return [condition, options];
+    },
+
+    parseOptions(options, engine) {
+        let returnOptions = {};
+        let whenOptions = {};
+
+        if (utils.isString(options)) {
+            whenOptions = parseSchemaString(options);
+        }
+        else {
+            // Copy, don't mutate.
+            whenOptions = { ...options };
+        }
+
+        // Only continue if we do have a valid object.
+        if (utils.isObject(whenOptions)) {
+            const keys = Object.keys(whenOptions);
+            for (let i = 0; i < keys.length; i += 1) {
+                const key = keys[i];
+
+                const validOptions = [ "is", "then", "otherwise" ];
+                if (validOptions.includes(key)) {
+                    returnOptions[key] = whenParser.parseValidOption(whenOptions[key], engine);
+                }
+            }
+        }
+
+        return returnOptions;
+    },
+
+    parseValidOption(options, engine) {
+        let returnValues = null;
+        // Convert array (alternatives).
+        if (utils.isArray(options)) {
+            returnValues = [];
+            for (let i = 0; i < options.length; i += 1) {
+                returnValues[i] = whenParser.parseValidOption(options[i], engine);
+            }
+        }
+        else if (utils.isObject(options)) {
+            let schema = engine;
```

Comment:
I wanted to clone this, but I didn't find a good way to deep clone. Since `engine` is essentially `Joi`, calls to Joi clone the object, it should be separating itself from the original as soon as the first call is made.

Reply:
`cloneDeep`: https://www.npmjs.com/package/clone-deep

Also available in Lodash.

Reply:
Do you think it's a good idea to add clone-deep? Depending on the shape of the condition, we could clone a few times, in each full whenParser.parse(), but most likely it will only happen once or twice

Reply:
I think it is unnecessary here.